### ISSUE TITLE
if parts are missing, don't make it impossible to try to download them later (except for livestreams)

### ISSUE THREAD

**Issue by chrizilla** [yt-dlp/yt-dlp#6793]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### problem
The current default settings (`--skip-unavailable-fragments`) skip missing parts and then concat the remaining available parts so that it is impossible to try to finish the missing parts later. Once the file is merged, the only way to get a complete video is a full redownload of the video.

### request
If fragments (part.files) of a download are missing for whatever reason, don't make it impossible by default to finish the unfinished fragments on a later yt-dlp rerun.

### rationale
The possible reasons for missing fragments are manifold and stretch from [network errors](https://github.com/yt-dlp/yt-dlp/issues/6078#issuecomment-1501307741) to [write errors](#6757).

### implemenation
I'll set this issue up as a place to collect and bring together the previously expressed ideas scattered across multiple issue threads. I'll leave further discussion about the best solution to the experts here.
   1. In the exceptional and rare case of a missing fragment, yt-dlp could exceptionally forcibly trigger `--keep-fragments` when missing a fragment so that it has a chance to finish missing parts on a rerun. This should override the `--no-keep-fragments` default, but NOT override if the user explicitly set `--no-keep-fragments`. The partfiles hould be kept (either in addition to or instead of the merged final file) as long as the download is not 100% completed due to missing parts.
  
   2.  @pukkandan had [this idea](https://github.com/yt-dlp/yt-dlp/issues/6757#issuecomment-1505005929) : _"We download all the available fragments, but don't actually merge if something is unavailable."_ (This may or may not be the same ideas as in (1) above).
 
   3. @Jules-A idea [here](https://github.com/yt-dlp/yt-dlp/issues/6757#issuecomment-1504359147): 
   > I'd like missing fragments to be tried to obtained even after all other have been downloaded. Although....
   > 
   > > Once you have skipped a fragment, further fragments may have been already downloaded and concatted. It makes no  sense to retry them again.
   > 
   > The current way files are built from fragments would need to be completely re-written in order for it to be possible.
   > […] it would be helpful over just aborting and trying another time: It would help downloading from unreliable sources where some fragments are missing at certain times, similar to incompletely seeded torrents. It would allow you to get all the ones that are available so next time you try you don't have to worry about the remaining fragments.

   I'd like to reiterate that missing/unfinished fragments are not only the result of "unreliable sources" as characterized above, but also the result of problems on the user side (WLAN fluctuation, unstable Internet connection, Windows crash, disk full, etc.).

### related issues:
* #3730 
* #6757
* #6793 
* #6078

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

log [here](https://github.com/yt-dlp/yt-dlp/issues/6757#issuecomment-1501349837)

**Comment by pukkandan** [comment]
Why do you keep opening new issues for the same thing?

**Comment by pukkandan** [comment]
ah, I meant to say to track it in #6078, as explained in https://github.com/yt-dlp/yt-dlp/issues/6757#issuecomment-1505005929. I hadn't even seen this new issue when I replied https://github.com/yt-dlp/yt-dlp/issues/6757#issuecomment-1506136100

**Comment by pukkandan** [comment]
Doesn't really matter, but okay
